app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt

```diff
@@ -34,7 +39,37 @@ open class FeedViewModel : ViewModel() {
     }
 
     fun refresh() {
-        //todo: fix the implementation
+        if (getState().isLoading) {
+            return
+        }
+        updateState { copy(isLoading = true)  }
+        refreshTaskDisposable = FeedApiService.feedApiService.getFeed()
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe({ feedResponse ->
+                updateState { copy(feedItems = feedResponse.templatesMetadata.toFeedItems())}
+                updateState { copy(isLoading = false)  }
+            },{ error ->
+                networkErrorEvent.postValue(Event(error.message.toString()))
+                updateState { copy(isLoading = false)  }
```

Comment:
1) Update first the state
2) then update the world with the error.
3) As a best practice never use postValue use always value on the main thread.